File: mira-hub/src/lib/citation-identity.ts
/**
 * Citation identity — the name a TECHNICIAN sees on a cited source (gate E-1).
 *
 * The product's whole claim is grounded answers with cited sources. The
 * 2026-09-07 mobile teardown found the claim rendering as:
 *
 *     FILE  nameplate-12ac8c22-018a-4104-a247-d81c37bdb292.txt   p. 1
 *
 * under the answer "Serial number 49849 is listed on the nameplate [1]." The
 * `[1]` marker, the kind label and the page number were all correct. The name
 * was a database key. A technician cannot tell which document that is, cannot
 * say it aloud to a colleague, and cannot check it — a citation in form only.
 *
 * Two halves live here because they fix the same defect at different ends:
 *
 *   `nameplateDocTitle`   — the PRODUCER. A derived document's stored filename
 *                           IS its citation label (retrieval reads
 *                           `metadata->>'filename'` into `ManualChunk.title`,
 *                           and that becomes `sourceTitle`), so a generated doc
 *                           must be given a name a human would choose.
 *   `humanCitationTitle`  — the READ path. Documents already stored under a
 *                           generated name keep it forever, and they are the
 *                           ones being cited today. The reader refuses to hand
 *                           a database key to a person.
 *
 * Uploaded PDFs never had this problem: their stored filename is the name the
 * technician gave the file, which is why real citations off a handset read
 * "…_End_Trucks_Owners_Manual (2).pdf p.7". Only DERIVED documents needed a
 * name chosen for them.
 */

const UUID_RE = /[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/i;

/** Just enough of a confirmed nameplate identity to name the document. */
export type NameplateIdentityFields = {
  manufacturer?: string | null;
  model?: string | null;
  catalogNumber?: string | null;
};

/**
 * The filename — and therefore the citation label — for a technician-confirmed
 * nameplate document.
 *
 * Safe to change from the old `nameplate-<photo-uuid>.txt`: the filename is a
 * label, not a key. Re-confirmation dedups on the canonical file hash, and the
 * origin photo is found by `fileId` via `findVisibleOriginSource`; neither
 * reads this string.
 *
 * Deterministic, like the bytes it names — same identity in, same name out — so
 * a replayed confirmation still reuses its document rather than minting one.
 *
 * Never invents a machine: with no confirmed make or model it falls back to the
 * notebook, then to the bare kind. It never emits a UUID at any branch.
 */
export function nameplateDocTitle(opts: {
  identity: NameplateIdentityFields;
  notebookName: string;
}): string {
  const model = (opts.identity.model || opts.identity.catalogNumber || "").trim();
  const machine = [(opts.identity.manufacturer ?? "").trim(), model]
    .filter(Boolean)
    .join(" ")
    .trim();
  if (machine) return `Nameplate — ${machine}.txt`;
  const notebook = opts.notebookName.trim();
  if (notebook) return `Nameplate — ${notebook}.txt`;
  return "Nameplate.txt";
}

/** Why a citation label is not the stored title. Kept out of the UI; the point
 *  is that the trace can tell these apart even when the screen cannot. */
export type CitationTitleReason = "stored" | "missing" | "generated_key";

export type CitationTitleResult = {
  title: string;
  reason: CitationTitleReason;
};

/**
 * Resolve the label for a cited source, and say WHY it is what it is.
 *
 * A floor that silently rewrote every unreadable title to "Attached document"
 * would collapse *"the title lookup broke"* and *"this document has no name"*
 * into one indistinguishable outcome — a green surface that cannot tell a
 * working lookup from a dead one. So the reason is returned rather than
 * inferred, and callers log it. Same discipline as an ungrounded answer having
 * to SAY it is ungrounded instead of merely omitting its citations.
 */
export function resolveCitationTitle(rawTitle: string | null | undefined): CitationTitleResult {
  const title = (rawTitle ?? "").trim();
  if (!title) return { title: "Attached document", reason: "missing" };
  if (!UUID_RE.test(title)) return { title, reason: "stored" };

  // A generated name from before `nameplateDocTitle`. Keep whatever human
  // prefix it carries ("nameplate-<uuid>.txt" -> "Nameplate"); never invent a
  // machine name that the stored title did not contain.
  const prefix = title.split(UUID_RE)[0].replace(/[-_\s]+$/, "").trim();
  const cleaned = prefix ? prefix.charAt(0).toUpperCase() + prefix.slice(1) : "Attached document";
  return { title: cleaned, reason: "generated_key" };
}

File: ux_acceptance/detectors/citation_identity.py
"""E-1 — a citation must name the document, not identify it.

Observed on a Pixel 9a, production 1.1.0, 2026-09-07:

    FILE   nameplate-12ac8c22-018a-4104-a247-d81c37bdb292.txt      p. 1

under the answer "Serial number 49849 is listed on the nameplate [1]." The
marker, the kind label and the page number are all correct. The *name* is a
database key. A technician cannot tell which manual that is, cannot say it
aloud to a colleague, and cannot check it — which is the whole commercial
claim ("diagnose with cited sources") rendered as an internal identifier.

This is deliberately a shape check, not a quality one. It cannot tell a good
title from a bad one; it can only tell that we showed an id where a name
belongs. That is exactly the part that needs no taste, so it can gate.
"""

from __future__ import annotations

from ..derive import CITATION_LABEL, children_of, citation_title, require
from ..snapshot import UUID_RE, Snapshot
from .base import Finding, Verdict

TEST_ID = "E-1"

#: Other machine-identifier shapes that are equally unreadable to a technician.
_BARE_ID_HINTS = (
    "urn:",
    "sha256:",
)


def _looks_like_an_identifier(label: str) -> str | None:
    """Return the offending substring, or None when the label reads as a name."""
    match = UUID_RE.search(label)
    if match:
        return match.group(0)
    lowered = label.lower()
    for hint in _BARE_ID_HINTS:
        if hint in lowered:
            return hint
    # A long unbroken hex run is an id even without UUID dashes.
    for token in label.replace("/", " ").replace("_", " ").replace(".", " ").split():
        if len(token) >= 24 and all(c in "0123456789abcdefABCDEF" for c in token):
            return token
    return None


def detect_citation_identity(snapshot: Snapshot) -> Finding:
    """FAIL when any visible citation label identifies rather than names."""
    labels = require(snapshot, CITATION_LABEL)

    offenders: list[str] = []
    for node in labels:
        text = citation_title(node, children_of(snapshot, node.id))
        if not text:
            offenders.append(f"{node.id}: citation label is empty")
            continue
        offending = _looks_like_an_identifier(text)
        if offending:
            offenders.append(f"{node.id}: {text!r} contains {offending!r}")

    if offenders:
        return Finding(
            detector="citation_identity",
            test_id=TEST_ID,
            verdict=Verdict.FAIL,
            summary=(
                f"{len(offenders)} of {len(labels)} citation label(s) show a machine "
                "identifier where a document name belongs"
            ),
            evidence=tuple(offenders),
        )

    return Finding(
        detector="citation_identity",
        test_id=TEST_ID,
        verdict=Verdict.PASS,
        summary=f"{len(labels)} citation label(s) name a document",
    )


detect_citation_identity.test_id = TEST_ID  # type: ignore[attr-defined]
